text/0080-datetime-extension.md

```diff
@@ -0,0 +1,484 @@
+forbid(
+  principal,
+  action == Action::"access",
+  resource is Document
+) when {
+  [1,7].contains(context.now.dayOfWeek)
+};
+```
+
+**Permit access to a special opportunity for persons born on Leap Day.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"redeem",
+  resource is Prize
+) when {
+  principal.birthDate.day == 29 &&
+  principal.birthDate.month == 2 &&
+  context.now.day == 29 && 
+  context.now.month == 2
+};
+```
+
+**Forbid access to EU resources after Brexit**
+
+```cedar
+forbid(
+  principal,
+  action,
+  resource
+) when {
+  context.now.timestamp > datetime("2020-01-31T23:00:00Z") &&
+  context.location.countryOfOrigin == 'GB' &&
+  resource.owner == 'EU'
+}
+```
+
+## Detailed design
+
+This RFC proposes supporting two new extension types: `datetime`, which represents a particular instant of time, up to millisecond accuracy, and `duration` which represents a duration of time.
+To construct and manipulate these types we will provide the functions listed below.
+All of this functionality will be hidden behind a `datetime` feature flag (analogous to the current decimal and IP extensions), allowing users to opt-out.
+
+### Instants of Time (`datetime`)
+
+The `datetime(string)` function constructs a datetime value. Like with other extension function constructors, strict validation requires `string` to be a string literal, although evaluation/authorization support any string-typed expression. The string must be of one of the forms, and regardless of the timezone offset is always normalized to UTC:
+
+- `"YYYY-MM-DD"` (date only)
+- `"YYYY-MM-DDThh:mm:ssZ"` (UTC)
+- `"YYYY-MM-DDThh:mm:ss.SSSZ"` (UTC with millisecond precision)
+- `"YYYY-MM-DDThh:mm:ss(+/-)hhmm"` (With timezone offset in hours and minutes)
+- `"YYYY-MM-DDThh:mm:ss.SSS(+/-)hhmm"` (With timezone offset in hours and minutes and millisecond precision)
+
+The `datetime` type does not provide a way for a policy author to create a `datetime` from a numeric timestamp. One of the readable formats listed above must be used, instead.
+
+Values of type `datetime` have the following methods:
+
+- `.offset(duration)` returns a new `datetime`, offset by duration.
+- `.toDate()` returns a new `datetime`, truncating to the day, such that printing the `datetime` would have `00:00:00` as the time.
+- `.toTime()` returns a new `duration`, removing the days, such that only milliseconds since `.toDate()` are left. This is equivalent to `DT - DT.toDate()`
+
+Values of type `datetime` can also be used with the `-` operator.
+
+- `DT1 - DT2` returns the difference between `DT1` and `DT2` as a `duration`. (The inverse of `-` is `DT1.offset(duration)`)
+
+An invariant for the `-` operator is that when `DT1` is before `DT2` the resulting duration is negative.
+
+`DT1` may be negative (_i.e._ milliseconds _before_ `1970-01-01T00:00:00Z`) requiring that `-` be computed specially:
+
+```
+if DT1 < 0 && DT2 >= 0 then
+    result = -(DT1 - DT2)
+else
+    result = DT1 - DT2
+```
+
```

Comment:
`DT1` cannot be negative. There is no "negative Jan 1, 1971." Rather the _representation_ of a DT as a number can be negative. This is an implementation detail that should not be here. Rather, it should go in the "Representation" subsection below.